src/algebra/big_operators.lean

```diff
@@ -421,6 +430,22 @@ lemma prod_eq_one {f : α → β} {s : finset α} (h : ∀x∈s, f x = 1) : s.pr
 calc s.prod f = s.prod (λx, 1) : finset.prod_congr rfl h
   ... = 1 : finset.prod_const_one
 
+/-- If we can partition a product into subsets that cancel out, then the whole product cancels. -/
+@[to_additive]
+lemma prod_cancels_of_partition_cancels [R : setoid α] [decidable_rel R.r]
```

Comment:
In your application, `R` is a rather ad hoc setoid instead of a completely canonical setoid that comes as a type class. And I expect this will be the case most of the time. Then maybe you should use `R` as a regular parameter instead of a typeclass parameter (but register it as a typeclass with `letI` in the proof).